// @ondc/automation-mock-runner/dist/lib/runners/node-runner.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeRunner = void 0;
const worker_threads_1 = require("worker_threads");
const code_validator_1 = require("../validators/code-validator");
const path = __importStar(require("path"));
class NodeRunner {
    constructor(options = {}) {
        this.worker = null;
        this.executionId = 0;
        this.pendingExecutions = new Map();
        this.workerPath = path.join(__dirname, "../../../public/node-worker.js");
        this.maxMemoryMB = options.maxMemoryMB || 128;
        this.initWorker();
    }
    initWorker() {
        if (this.worker) {
            this.worker.terminate();
        }
        // Create worker with resource limits
        this.worker = new worker_threads_1.Worker(this.workerPath, {
            resourceLimits: {
                maxOldGenerationSizeMb: this.maxMemoryMB,
                maxYoungGenerationSizeMb: this.maxMemoryMB / 2,
                codeRangeSizeMb: 16,
            },
            // Prevent worker from accessing parent environment variables
            env: {},
        });
        this.worker.on("message", (message) => {
            const { id, success, result, error, logs, executionTime } = message;
            const pending = this.pendingExecutions.get(id);
            if (pending) {
                clearTimeout(pending.timeout);
                this.pendingExecutions.delete(id);
                pending.resolve({
                    timestamp: new Date().toISOString(),
                    success,
                    result,
                    error,
                    logs,
                    executionTime,
                    validation: { isValid: true, errors: [], warnings: [] },
                });
            }
        });
        this.worker.on("error", (error) => {
            console.error("Worker error:", error);
            // Resolve all pending executions with error
            this.pendingExecutions.forEach((pending, id) => {
                clearTimeout(pending.timeout);
                pending.resolve({
                    success: false,
                    timestamp: new Date().toISOString(),
                    error: {
                        message: error.message || "Worker crashed",
                        name: "WorkerError",
                    },
                    logs: [],
                    validation: { isValid: true, errors: [], warnings: [] },
                });
            });
            this.pendingExecutions.clear();
            this.initWorker();
        });
        this.worker.on("exit", (code) => {
            if (code !== 0) {
                console.error(`Worker stopped with exit code ${code}`);
                // Handle any pending executions
                this.pendingExecutions.forEach((pending, id) => {
                    clearTimeout(pending.timeout);
                    pending.resolve({
                        success: false,
                        timestamp: new Date().toISOString(),
                        error: {
                            message: `Worker exited with code ${code}`,
                            name: "WorkerExitError",
                        },
                        logs: [],
                        validation: { isValid: true, errors: [], warnings: [] },
                    });
                });
                this.pendingExecutions.clear();
            }
        });
    }
    async execute(functionBody, schema, args) {
        const validation = code_validator_1.CodeValidator.validate(functionBody, schema);
        if (!validation.isValid || !validation.wrappedCode) {
            return {
                success: false,
                timestamp: new Date().toISOString(),
                error: {
                    message: validation.errors.join("\n"),
                    name: "ValidationError",
                },
                logs: [],
                validation,
            };
        }
        return new Promise((resolve) => {
            const id = ++this.executionId;
            const timeout = schema.timeout || 5000;
            const timeoutId = setTimeout(() => {
                this.pendingExecutions.delete(id);
                resolve({
                    success: false,
                    timestamp: new Date().toISOString(),
                    error: {
                        message: `Execution timeout after ${timeout}ms`,
                        name: "TimeoutError",
                    },
                    logs: [],
                    validation,
                });
                // Restart worker to kill the hanging execution
                this.initWorker();
            }, timeout);
            this.pendingExecutions.set(id, {
                resolve: (result) => {
                    resolve({ ...result, validation });
                },
                timeout: timeoutId,
            });
            // Send code to worker for execution
            this.worker?.postMessage({
                id,
                code: validation.wrappedCode,
                functionName: schema.name,
                args,
                timeout,
            });
        });
    }
    terminate() {
        // Clear all pending executions
        this.pendingExecutions.forEach((pending) => {
            clearTimeout(pending.timeout);
            pending.resolve({
                success: false,
                timestamp: new Date().toISOString(),
                error: {
                    message: "Runner terminated",
                    name: "TerminationError",
                },
                logs: [],
                validation: { isValid: true, errors: [], warnings: [] },
            });
        });
        this.pendingExecutions.clear();
        // Terminate worker
        this.worker?.terminate();
        this.worker = null;
    }
}
exports.NodeRunner = NodeRunner;
