// @ondc/automation-mock-runner/dist/lib/types/mock-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockPlaygroundConfigSchema = exports.TransactionHistoryItemSchema = exports.PlaygroundActionStepSchema = exports.MockConfigSchema = exports.TransactionDataSchema = exports.MetaSchema = exports.OwnerSchema = void 0;
const zod_1 = require("zod");
// Zod schemas
exports.OwnerSchema = zod_1.z.enum(["BAP", "BPP"]);
exports.MetaSchema = zod_1.z.object({
    domain: zod_1.z.string().min(1, "Domain is required"),
    version: zod_1.z.string().min(1, "Version is required"),
    flowId: zod_1.z.string().min(1, "Flow ID is required"),
});
exports.TransactionDataSchema = zod_1.z.object({
    transaction_id: zod_1.z.string().min(1, "Transaction ID is required"),
    latest_timestamp: zod_1.z.string().min(1, "Latest timestamp is required"),
    bap_id: zod_1.z.string().optional(),
    bap_uri: zod_1.z.string().min(1, "BAP URI is required").optional(),
    bpp_id: zod_1.z.string().min(1, "BPP ID is required").optional(),
    bpp_uri: zod_1.z.string().min(1, "BPP URI is required").optional(),
});
exports.MockConfigSchema = zod_1.z.object({
    generate: zod_1.z.base64(),
    validate: zod_1.z.base64(),
    requirements: zod_1.z.base64(),
    defaultPayload: zod_1.z.any(),
    saveData: zod_1.z.record(zod_1.z.string(), zod_1.z.string()),
    inputs: zod_1.z.object({
        id: zod_1.z.string().min(1, "Input ID is required").optional(),
        jsonSchema: zod_1.z.any().optional(),
    }),
});
exports.PlaygroundActionStepSchema = zod_1.z.object({
    api: zod_1.z.string().min(1, "API is required"),
    action_id: zod_1.z.string().min(1, "Action ID is required"),
    owner: exports.OwnerSchema,
    responseFor: zod_1.z.string().nullable(),
    unsolicited: zod_1.z.boolean(),
    description: zod_1.z.string(),
    repeatCount: zod_1.z
        .number()
        .min(1, "Repeat count must be at least 1")
        .optional()
        .nullable(),
    mock: exports.MockConfigSchema,
});
exports.TransactionHistoryItemSchema = zod_1.z.object({
    action_id: zod_1.z.string().min(1, "Action ID is required"),
    payload: zod_1.z.any(),
    saved_info: zod_1.z.record(zod_1.z.string(), zod_1.z.any()),
});
exports.MockPlaygroundConfigSchema = zod_1.z.object({
    meta: exports.MetaSchema,
    transaction_data: exports.TransactionDataSchema,
    steps: zod_1.z.array(exports.PlaygroundActionStepSchema),
    transaction_history: zod_1.z.array(exports.TransactionHistoryItemSchema),
    validationLib: zod_1.z.string(),
    helperLib: zod_1.z.string(),
});
