// @ondc/automation-mock-runner/dist/lib/utils/errors.js
"use strict";
/**
 * Custom error classes for better error handling and debugging
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionDataError = exports.ActionNotFoundError = exports.TimeoutError = exports.SecurityError = exports.ExecutionError = exports.ValidationError = exports.ConfigurationError = exports.MockRunnerError = void 0;
class MockRunnerError extends Error {
    constructor(message, code, context = {}) {
        super(message);
        this.name = this.constructor.name;
        this.code = code;
        this.context = context;
        this.timestamp = new Date().toISOString();
        // Capture stack trace
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, this.constructor);
        }
    }
    toJSON() {
        return {
            name: this.name,
            message: this.message,
            code: this.code,
            context: this.context,
            timestamp: this.timestamp,
            stack: this.stack
        };
    }
}
exports.MockRunnerError = MockRunnerError;
class ConfigurationError extends MockRunnerError {
    constructor(message, context = {}) {
        super(message, 'CONFIG_ERROR', context);
    }
}
exports.ConfigurationError = ConfigurationError;
class ValidationError extends MockRunnerError {
    constructor(message, validationErrors = [], context = {}) {
        super(message, 'VALIDATION_ERROR', context);
        this.validationErrors = validationErrors;
    }
}
exports.ValidationError = ValidationError;
class ExecutionError extends MockRunnerError {
    constructor(message, actionId, executionTime = 0, context = {}) {
        super(message, 'EXECUTION_ERROR', { ...context, actionId, executionTime });
        this.executionTime = executionTime;
        this.actionId = actionId;
    }
}
exports.ExecutionError = ExecutionError;
class SecurityError extends MockRunnerError {
    constructor(message, violation, context = {}) {
        super(message, 'SECURITY_ERROR', { ...context, violation });
        this.securityViolation = violation;
    }
}
exports.SecurityError = SecurityError;
class TimeoutError extends MockRunnerError {
    constructor(message, timeoutMs, context = {}) {
        super(message, 'TIMEOUT_ERROR', { ...context, timeoutMs });
        this.timeoutMs = timeoutMs;
    }
}
exports.TimeoutError = TimeoutError;
class ActionNotFoundError extends MockRunnerError {
    constructor(actionId, availableActions = []) {
        super(`Action step with ID '${actionId}' not found`, 'ACTION_NOT_FOUND', {
            actionId,
            availableActions
        });
        this.actionId = actionId;
        this.availableActions = availableActions;
    }
}
exports.ActionNotFoundError = ActionNotFoundError;
class SessionDataError extends MockRunnerError {
    constructor(message, stepIndex, jsonPath, context = {}) {
        super(message, 'SESSION_DATA_ERROR', { ...context, stepIndex, jsonPath });
        this.stepIndex = stepIndex;
        this.jsonPath = jsonPath || '';
    }
}
exports.SessionDataError = SessionDataError;
