// @ondc/automation-mock-runner/dist/lib/utils/logger.js
"use strict";
/**
 * Enhanced logging utility for MockRunner
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = exports.LogLevel = void 0;
var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["ERROR"] = 0] = "ERROR";
    LogLevel[LogLevel["WARN"] = 1] = "WARN";
    LogLevel[LogLevel["INFO"] = 2] = "INFO";
    LogLevel[LogLevel["DEBUG"] = 3] = "DEBUG";
})(LogLevel || (exports.LogLevel = LogLevel = {}));
class Logger {
    constructor() {
        this.logLevel = LogLevel.INFO;
        this.logs = [];
        this.maxLogEntries = 1000;
    }
    static getInstance() {
        if (!Logger.instance) {
            Logger.instance = new Logger();
        }
        return Logger.instance;
    }
    setLogLevel(level) {
        this.logLevel = level;
    }
    shouldLog(level) {
        return level <= this.logLevel;
    }
    addLog(level, message, context, error) {
        if (!this.shouldLog(level))
            return;
        const entry = {
            timestamp: new Date().toISOString(),
            level,
            message,
            context,
            error,
        };
        this.logs.push(entry);
        // Keep only recent logs
        if (this.logs.length > this.maxLogEntries) {
            this.logs = this.logs.slice(-this.maxLogEntries);
        }
        // Also log to console in development
        if (process.env.NODE_ENV === 'development' || process.env.DEBUG) {
            const levelNames = ['ERROR', 'WARN', 'INFO', 'DEBUG'];
            console.log(`[${entry.timestamp}] ${levelNames[level]}: ${message}`, context || '');
            if (error)
                console.error(error);
        }
    }
    error(message, context, error) {
        this.addLog(LogLevel.ERROR, message, context, error);
    }
    warn(message, context) {
        this.addLog(LogLevel.WARN, message, context);
    }
    info(message, context) {
        this.addLog(LogLevel.INFO, message, context);
    }
    debug(message, context) {
        this.addLog(LogLevel.DEBUG, message, context);
    }
    getLogs(level) {
        if (level !== undefined) {
            return this.logs.filter(log => log.level <= level);
        }
        return [...this.logs];
    }
    clearLogs() {
        this.logs = [];
    }
    // Helper for execution tracking
    createExecutionContext(actionId) {
        const executionId = `${actionId}-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
        this.debug(`Starting execution`, { actionId, executionId });
        return executionId;
    }
    logExecution(executionId, message, context) {
        this.info(message, { ...context, executionId });
    }
}
exports.Logger = Logger;
