// @ondc/automation-mock-runner/dist/test/__mocks__/browser-globals.js
"use strict";
/**
 * Mock browser environment for testing BrowserRunner
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.mockBrowserGlobals = void 0;
// Mock browser globals
const mockBrowserGlobals = () => {
    // Mock Worker class
    class MockWorker extends EventTarget {
        constructor(url) {
            super();
            this.messageHandlers = new Set();
            // Simulate worker creation delay
            setTimeout(() => {
                this.dispatchEvent(new Event("load"));
            }, 10);
        }
        postMessage(data) {
            // Simulate worker message processing
            setTimeout(async () => {
                const startTime = Date.now();
                try {
                    const { id, code, functionName, args } = data;
                    // Execute the code in current context (simulating worker)
                    const logs = [];
                    // Mock console for capturing logs
                    const mockConsole = {
                        log: (...logArgs) => logs.push(logArgs.join(" ")),
                        error: (...logArgs) => logs.push(logArgs.join(" ")),
                        warn: (...logArgs) => logs.push(logArgs.join(" ")),
                    };
                    // Create execution context
                    const context = {
                        console: mockConsole,
                        // Add other safe globals as needed
                        Math,
                        Date,
                        JSON,
                        String,
                        Number,
                        Boolean,
                        Array,
                        Object,
                        Promise,
                        setTimeout,
                        clearTimeout,
                    };
                    // Execute code with mock context
                    const fn = new Function(...Object.keys(context), code + `; return ${functionName};`);
                    const userFunction = fn(...Object.values(context));
                    if (typeof userFunction !== "function") {
                        throw new Error(`${functionName} is not a function`);
                    }
                    // Handle both sync and async functions
                    let result = userFunction(...args);
                    // If result is a promise, await it
                    if (result && typeof result.then === "function") {
                        result = await result;
                    }
                    const executionTime = Date.now() - startTime;
                    // Simulate async response
                    this.dispatchEvent(new MessageEvent("message", {
                        data: {
                            id,
                            success: true,
                            result,
                            logs,
                            executionTime,
                        },
                    }));
                }
                catch (error) {
                    const executionTime = Date.now() - startTime;
                    this.dispatchEvent(new MessageEvent("message", {
                        data: {
                            id: data.id,
                            success: false,
                            error: {
                                message: error.message,
                                name: error.name,
                                stack: error.stack,
                            },
                            logs: [],
                            executionTime,
                        },
                    }));
                }
            }, 50); // Simulate network delay
        }
        terminate() {
            // Clean up
            this.messageHandlers.clear();
        }
        addEventListener(type, listener) {
            super.addEventListener(type, listener);
        }
        removeEventListener(type, listener) {
            super.removeEventListener(type, listener);
        }
    }
    // Mock Blob for worker URL creation
    class MockBlob {
        constructor(content, options = {}) {
            this.content = content;
            this.options = options;
        }
    }
    // Mock URL for createObjectURL
    const MockURL = {
        createObjectURL: (blob) => {
            return "blob:mock-url-" + Math.random().toString(36).substr(2, 9);
        },
        revokeObjectURL: (url) => {
            // Mock cleanup
        },
    };
    return {
        Worker: MockWorker,
        Blob: MockBlob,
        URL: MockURL,
        window: {}, // Mock window object
        document: {}, // Mock document object
        performance: {
            now: () => Date.now(),
        },
    };
};
exports.mockBrowserGlobals = mockBrowserGlobals;
